### PR TITLE
Traefik Pilot: plugins support and alert system (EXPERIMENTAL FEATURES)

### LINKED ISSUE #1336: enable custom plugins/middlewares for Traefik
After seeing the [Go1.8 new plugin feature](https://golang.org/pkg/plugin/) I though that this could help a lot o people to add specific functionalities to Traefik.

Instead of building/compiling/shipping a custom-made version of Traefik to enable a custom functionality it would be possible to write way simpler custom-made middlewares with this approach, doesn't it ? 

Try imagine creating a package that receives the request at a parameter without having to recompile the whole Traefik repository just to add a small change. Does it sounds like a middleware ? Because for me it is ! It's just a go1.8-plugin-based-middleware !

What do you guys think ?

### PR DESCRIPTION
**Pull request by ldez**
### What does this PR do?

We are proud to introduce two new major features :heart: :

- Traefik Pilot: Traefik Pilot is a software-as-a-service (SaaS) platform that connects to Traefik to extend its capabilities.
- Plugins support (middleware): to extend Traefik capabilities (Traefik Pilot is used to handle plugins)

**EXPERIMENTAL FEATURES**

### Motivation

Fixes #1336

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

This PR contains the documentation about plugins but you can also take a look at the following repository to have explanations about how to create a plugin: https://github.com/containous/plugindemo

### CODE REVIEW

File: pkg/plugins/builder.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+			Import:      manifest.Import,
+			BasePkg:     manifest.BasePkg,
+		}
+	}
+
+	return pb, nil
+}
+
+// Build builds a plugin.
+func (p Builder) Build(pName string, config map[string]interface{}, middlewareName string) (*Middleware, error) {
```
Comment (mmatur) [inline pkg/plugins/builder.go:92]: Could you rename `p` into `b`

File: pkg/plugins/builder.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+	return &Middleware{
+		middlewareName: middlewareName,
+		fnNew:          fnNew,
+		config:         vConfig,
+	}, nil
+}
+
+// NewHandler creates a new HTTP handler.
+func (w *Middleware) NewHandler(ctx context.Context, next http.Handler) (http.Handler, error) {
```
Comment (mmatur) [inline pkg/plugins/builder.go:152]: Could you rename `w` into `m`

File: pkg/plugins/client.go [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+
+// ClientOptions the options of a Traefik Pilot client.
+type ClientOptions struct {
+	Output string
+	Token  string
+}
+
+// Client a Traefik Pilot client.
+type Client struct {
+	HTTPClient *http.Client
```
Comment (mmatur) [inline pkg/plugins/client.go:47]: Is there a reason to have this client exposed?

File: pkg/plugins/client.go [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+	if resp.StatusCode == http.StatusNotModified {
+		// noop
+		return hash, nil
+	}
+
+	data, _ := ioutil.ReadAll(resp.Body)
+	return "", fmt.Errorf("error: %d: %s", resp.StatusCode, string(data))
+}
+
+// Check check the plugin archive integrity.
```
Comment (mmatur) [inline pkg/plugins/client.go:204]:
```suggestion
// Check checks the plugin archive integrity.
```

File: pkg/plugins/client.go (hunk 3 above)
Comment (ldez) [reply pkg/plugins/client.go:47]: yes, it allow to override the HTTP client for testing.

**Comment by Julio-Guerra** [comment]
Congratulations on this great achievement.
Quick feedback after giving it a quick try: it seems to be still not enough for advanced security middlewares involving cgo and some low-level Go. Yaegi doesn't seem to support it.

**Pull request merged**